### DISCUSSION TITLE
Development branch: Little Backup Box core files are nearly totally rewritten to python

### DISCUSSION THREAD

**Discussion by rkosick** [outdoorbits/little-backup-box#232] [General]
I installed the new development branch on a clean system on RP Zero 2 W. Unfortunatelly it seems to stuck on the splash screen.
Log file only shows:

Little Backup Box
2023-08-30 18:52:09
IMAGE=/var/www/little-backup-box/little-backup-box.bmp

Connection to http://10.41.0.1:8000 I can manage settings. But things like reboot or power down do seem to do nothing.

**Comment by outdoorbits** [comment]
Is this a brand new installation? I uploaded a lot of bug fixes, power commands included. Can you start backups?
If the installation is not from just now,  please just run the installer command again.
` curl -sSL https://raw.githubusercontent.com/outdoorbits/little-backup-box/development/install-little-backup-box.sh | bash -s -- development 2> install-error.log`

**Reply by rkosick** [reply]
Yes it was a brand new installation. Tried to start the backup via the web interface. This was not possible because I didn't set a global password yet. After that and the update, lbb is working now. Not sure if the missing global password was the problem. Before that I did again run the installer command and it didn't do anything resolving the issue. No increase in copy speed on the raspberry pi zero 2 w, Still stuck at 17 MB/s but thats absolute ok with me. Maybe this is also regarding my ssd. I will test the other things I normally use the next days. Thanks for your great work.

**Comment by outdoorbits** [comment]
Your version might have 'updated' to the main branch. The very last version in development has an option to update development also.
If your backup page has two columns of buttons, each defining source and target, it is the old version.

**Reply by rkosick** [reply]
This sounds strange. I think I never had two columns before. At least I only saw that while trying upgrade to development branch for one time today. Still have the old sd card and there is also only one colum. So that must be really old version with 2 columns. Never mind, it works now.

**Comment by outdoorbits** [comment]
Else just check in settings- update. If there are two options, one to update to development, you are in development version.

**Reply by rkosick** [reply]
Ok. looks I'm not. I'll have a look into tomorrow. Thanks.

**Comment by rkosick** [comment]
Still problems with installation of development version. After new setup lbb is still stuck on spash screen. Interface available but restart and reboot is not working still. Again I set a global password and used the updater again for development version. Saw this option now. But this time it is not fixing the problem. Something I saw after initial installation: Installer says that it would reboot the pi but doesn't do it anyway. Even after updating it via web interface it doesn't. 
[install-error.log](https://github.com/outdoorbits/little-backup-box/files/12488829/install-error.log)

See also attachted install-error.log

**Reply by rkosick** [reply]
Is "Raspberry Pi OS Lite (32-bit)" or "Raspberry Pi OS Lite (64-bit)" still right? Not the legacy option? I was using the 64 Bit version.

**Comment by outdoorbits** [comment]
I think I got it! Your RP Zero 2 W has no LEDs, right? It crashes controlling them... Bugfix soon!

**Comment by outdoorbits** [comment]
Please, update again, the LED bug is fixed for pi zero.

**Comment by rkosick** [comment]
Still not working. It reboots now after installation but interface looks not right. I don't see the IP address and backup is not working. See install-error.log
Also tried the update over the interface. First it stopped while cloning Little backup box and the second time it just lost connection.
[install-error.log](https://github.com/outdoorbits/little-backup-box/files/12499143/install-error.log)
![Screenshot_20230901-163855](https://github.com/outdoorbits/little-backup-box/assets/122625940/85cee616-71e3-4ecb-92bc-c899f10eaad6)
![Screenshot_20230901-163243](https://github.com/outdoorbits/little-backup-box/assets/122625940/ab2da8e4-d6c8-499c-bc6b-1dfa2c46aa8a)

**Comment by outdoorbits** [comment]
Looks close to good. I didn't see any installation problems. But CSS of the web ui has changed and your browser obviously cached it. Just do a Reload (Override Cache) on your browser, for example on Firefox CTRL-F5. Now select a source and start process by klicking on the target.

**Reply by rkosick** [reply]
Like I said, it's not only the the interface. Backup is not working and also IP adress is not showing. Interface in browser was now ok, started new update but didn't help.

**Comment by outdoorbits** [comment]
Then my assessment was probably too euphoric. But step by step. If you don't lose patience, neither will I.
Could you please update again, login via terminal/ssh and try one of these backups:

`sudo python3 /var/www/little-backup-box/backup.py usb usb` 
or
`sudo python3 /var/www/little-backup-box/backup.py camera usb`

and tell me the response on terminal?

**Reply by rkosick** [reply]
Thanks for your patience. Without doing a clean install this is the result:

pi@raspberrypi:~ $ sudo python3 /var/www/little-backup-box/backup.py usb usb
umount: /media/target_local: no mount point specified.
umount: /media/source_local: no mount point specified.
umount: /mnt/target_local: no mount point specified.
umount: /mnt/source_local: no mount point specified.
umount: /media/target_cloud: no mount point specified.
umount: /media/source_cloud: no mount point specified.

Not reacting anymore so I did CTRL+C:

^CTraceback (most recent call last):
  File "/var/www/little-backup-box/backup.py", line 1057, in <module>
    backupObj= backup(SourceName, TargetName, DoSyncDatabase, DoGenerateThumbnails, DoUpdateEXIF, DeviceIdentifierPresetSource, DeviceIdentifierPresetTarget, PowerOff, SecundaryBackupFollows)
  File "/var/www/little-backup-box/backup.py", line 252, in __init__
    self.TargetDevice.mount()
  File "/var/www/little-backup-box/lib_storage.py", line 209, in mount
    mounted	= self.__mount_USB_storage()
  File "/var/www/little-backup-box/lib_storage.py", line 250, in __mount_USB_storage
    subprocess.run(['sudo','udevadm','trigger'])
  File "/usr/lib/python3.9/subprocess.py", line 507, in run
    stdout, stderr = process.communicate(input, timeout=timeout)
  File "/usr/lib/python3.9/subprocess.py", line 1126, in communicate
    self.wait()
  File "/usr/lib/python3.9/subprocess.py", line 1189, in wait
    return self._wait(timeout=timeout)
  File "/usr/lib/python3.9/subprocess.py", line 1919, in _wait
    (pid, sts) = self._try_wait(0)
  File "/usr/lib/python3.9/subprocess.py", line 1877, in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
KeyboardInterrupt

AND:

pi@raspberrypi:~ $ sudo python3 /var/www/little-backup-box/backup.py usb usb
umount: /media/target_local: no mount point specified.
umount: /media/source_local: no mount point specified.
umount: /mnt/target_local: no mount point specified.
umount: /mnt/source_local: no mount point specified.
umount: /media/target_cloud: no mount point specified.
umount: /media/source_cloud: no mount point specified.

Not reacting anymore so I did CTRL+C:

^C^CTraceback (most recent call last):
  File "/var/www/little-backup-box/backup.py", line 1057, in <module>
    backupObj= backup(SourceName, TargetName, DoSyncDatabase, DoGenerateThumbnails, DoUpdateEXIF, DeviceIdentifierPresetSource, DeviceIdentifierPresetTarget, PowerOff, SecundaryBackupFollows)
  File "/var/www/little-backup-box/backup.py", line 252, in __init__
    self.TargetDevice.mount()
  File "/var/www/little-backup-box/lib_storage.py", line 209, in mount
    mounted	= self.__mount_USB_storage()
  File "/var/www/little-backup-box/lib_storage.py", line 250, in __mount_USB_storage
    subprocess.run(['sudo','udevadm','trigger'])
  File "/usr/lib/python3.9/subprocess.py", line 507, in run
    stdout, stderr = process.communicate(input, timeout=timeout)
  File "/usr/lib/python3.9/subprocess.py", line 1126, in communicate
    self.wait()
  File "/usr/lib/python3.9/subprocess.py", line 1189, in wait
    return self._wait(timeout=timeout)
  File "/usr/lib/python3.9/subprocess.py", line 1919, in _wait
    (pid, sts) = self._try_wait(0)
  File "/usr/lib/python3.9/subprocess.py", line 1877, in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
KeyboardInterrupt

**Comment by outdoorbits** [comment]
If you run
sudo udevadm trigger
Do you get an error?

I found some stupid errors in unmount process, do please update to the last version.

**Reply by rkosick** [reply]
I did not test sudo udevadm trigger because I tried to update right away. I can say it doesn't show errors now. Still not working as expected.
After fresh installation I get the following error log:
[install-error.log](https://github.com/outdoorbits/little-backup-box/files/12504553/install-error.log)

Still not getting an IP adress on the display. Web interface is working but slow (see later regarding CPU utilization).
Starting a backup from usb to usb the display is not refreshing. It stucks at copying the first file but it does indeed copy all the files but is not showing it. See log:
[little-backup-box.log](https://github.com/outdoorbits/little-backup-box/files/12504558/little-backup-box.log)

Could be for CPU utilization that is now at more than 80%. htop shows that /usr/bin/python3 /var/www/little-backup-box/display.py is at more than 72% utiliziation. On a ARM Cortex-A53 Quad-Core on Pi Zero 2 W. I have no issue with the stable version regading CPU utilization. 

I tried update over web interface but again, it stucks at "Cloning Little backup box"

**Comment by outdoorbits** [comment]
There was a bug displaying the IP, it's fixed now.
About the progress I have not really an idea yet. But maybe we start at the high CPU utilisation. Can your find out the process taking all that power?

**Reply by rkosick** [reply]
Like I said htop shows that /usr/bin/python3 /var/www/little-backup-box/display.py is at more than 72% utiliziation. Not sure how to get more details now.

**Comment by outdoorbits** [comment]
Bug fixed, CPU load should be as low as before. What's about the progress display, is this affected by lower CPU load?

**Reply by rkosick** [reply]
CPU load is now much better. Still high after start regarding /lib/systemd/systemd-udev running multiple instances but after starting the backup process CPU utilization is down like before. IP adress in display is working also. Great!
Progress on display still shows only the first file for at least a minute. Then it startet updating showing the progress of up to 9 files copied (with a remaining time that makes no sense - about 2,6 GB of data with transfer rate of more than 17MB/s shows 35 min remaining time) and then I get an error trying another backup until it stops. But all the files were copied.
Also: There is no ID.file created on the source.
See attached log:
[little-backup-box.log](https://github.com/outdoorbits/little-backup-box/files/12507411/little-backup-box.log)

**Reply by rkosick** [reply]
Regarding time stemps in the log file: There is some time between mounting the target and source. Thats was not a problem, I just needed some time before connection the source because I deleted pictures so I would get a result in shorter time.

**Reply by rkosick** [reply]
And also: updating via web interface now works fine.

**Reply by rkosick** [reply]
After updating via web interface the progress is showing from the beginning. I added files and it starts copying all files again in a new folder. But I get an error again "lost source device usb", maybe because of a power issue?:
[little-backup-box.log](https://github.com/outdoorbits/little-backup-box/files/12507584/little-backup-box.log)

**Comment by outdoorbits** [comment]
Maybe there was an error calculating files to sync. That is fixed.

**Comment by outdoorbits** [comment]
ID file is fixed. Estimated time is calculated from jobs done, jobs in total and elapsed time while transfer speed is a moment feedback of rsync.
It's still the question, why you don't get the progress earlier. As the progress is counted, the information comes but late, it's like buffered - but buffering is disabled.

**Reply by rkosick** [reply]
Updating via web interface damaged the installation. Did try a new installation, again no reboot at the end. And its not working at all. Display stays dark, log monitor not showing anything. CPU utilizsation is at zero.

[install-error.log](https://github.com/outdoorbits/little-backup-box/files/12515828/install-error.log)

**Comment by outdoorbits** [comment]
Ups, I forgot to remove a local test config. That broke the installation. It's fixed.

**Reply by rkosick** [reply]
Installation works now again. Backup is working but still strange behavior with the output.
First inital backup of about 100 files starts normal and is not updating after a few files. Again all files are copied anyways.

[little-backup-box_first_copy.log](https://github.com/outdoorbits/little-backup-box/files/12525485/little-backup-box_first_copy.log)

A second attempt with another new more over 50 files is calculated correct and is also copying the files correectly, but it says it missed 3 files. In reality, only one read only file named SONYCARD.IND is not copied. Again output is not updated over the hole process.

CPU utilization is not high, so this should not be a problem. I get a good transfer speed. Maybe there is another bottleneck to it.

[little-backup-box_second_copy.log](https://github.com/outdoorbits/little-backup-box/files/12525552/little-backup-box_second_copy.log)

**Reply by outdoorbits** [reply]
It's this Sony file a real file? Can you open or copy it? If I'm not mistaken, Sony partitions have their own quirks. For example, they don't have a uuid. Perhaps this file is such a special feature? Maybe this file should just be excluded? Does it have any function  outside the camera?
Today I don't have the time to read up ...

Progress monitoring is something different.   Progress is counted correctly but - why ever - not monitored. You have to know, there is a frame time defined. Every page is displayed at least this time. Next progress step is also only displayed not before this time elapsed. There are two exceptions: 0% and 100% are always shown. It's a miracle to me why this is not working in your configuration. But we'll find out...

**Reply by outdoorbits** [reply]
Your log files: Did you copy regular image files or maybe very small test files where 50 files could be copied in seconds?

**Reply by rkosick** [reply]
It is a real files and I can copy it and open it. But it is empty and doesn't have a purpose I guess.

I copy regular Files that are about 10 MB each so it takes almost 2 minutes.

**Comment by rkosick** [comment]
Testet the last stable version again and this is working like before. So no hardware issue.

[little-backup-box_stable.log](https://github.com/outdoorbits/little-backup-box/files/12527267/little-backup-box_stable.log)

**Comment by outdoorbits** [comment]
Great. I hope I found to solution of the progress problem, but it's too late today to test and upload.
The problem of the missing empty file is easy: There was an issue of (false) empty files on target. So I did a) exclude files of 0 byte from sync and after sync remove empty files - they must be corrupted. But an empty file at source should not raise an error.

**Reply by rkosick** [reply]
No hurry.

**Reply by rkosick** [reply]
Btw: Are you including these updates like the one with empty files beeing ignored, in the stable version too?

**Comment by rkosick** [comment]
Btw: One thing I always changed in the initial installation: I undervolt the CPU a little bit and underclock it from 1GHz to 800 MHz. Because I wanted it as compact as possible without a fan. Doing so, I don't run in throtteling in hot environments. Not sure if this would be a problem regarding the new development version of lbb. It's part of my install routine and I would be ashamed if this would be the problem all the time. I think throtteling was a major problem in a ancient version of the project.

**Comment by outdoorbits** [comment]
Thank you, but I don't think that's the problem.
I still have a little hope being able to upload tomorrow, but actually it seems to be more and more unlikely. Most probably monday I'll upload and I'm very excited if it's working in your setting!

**Comment by outdoorbits** [comment]
It is already part of stable version. Do you think that's a disadvantage?

**Reply by rkosick** [reply]
No thats great. Just wanted to make sure that things like that are not only considered for the development branch. I think there was also something regarding not using partial file transfer option on local backups using rsync and stuff like that. Happy to hear that. I currently have 2 versions on SD cards as long as the development version is not running perfectly for me.

**Reply by rkosick** [reply]
It also fixes some confusing output at the end, on a canon camera of a friend.

**Comment by outdoorbits** [comment]
There has been a lot to do but now progress is running fine in my installation. There are many more bugfixes. VPN is working, just tested it using WireGuard. If you configured a auto backup, please check your setup and save again.

**Reply by rkosick** [reply]
Did also try a new installation but still have the same with progress.
[little-backup-box (16).log](https://github.com/outdoorbits/little-backup-box/files/12588073/little-backup-box.16.log)

In addition the config file doesn't get saved. Therefore after backup thumbnail creation was still active. Tried a few times to save the config with different buttons but didn't work. At least for the thumbnails. Saving the global password did work because I was asked for the password to connect to the wifi ap.

**Comment by outdoorbits** [comment]
An error in file ownership, fixed...

**Reply by outdoorbits** [reply]
From 158 files there are 150 selected to create thumbnails. Do you know, what file types the missing ones are? Maybe tif or other not yet supported types?
Are the file names or path names somehow special?
I hope the problem comes from a double check to be sure the file is really transferred. Are there special characters in names or are there names starting or ending with spaces or something like that?

**Reply by rkosick** [reply]
150 thumbnails is correct. There where 150 pictures. The rest is random stuff from the camera
The Path is pretty normal. /DCIM/100MSDCF/ and files also short: _DSC4133.JPG
So there is a special character starting the filenames.

**Comment by outdoorbits** [comment]
I hope, that's it. I'll remove that double check and ask you to test again then...

**Reply by rkosick** [reply]
Just tell me when it's done. No hurry and thanks for your patience with me.

**Comment by outdoorbits** [comment]
would you please check again...

**Reply by rkosick** [reply]
Did a quick test. Now it looks great. Needed a new installation because global password was also not saved. So I couldn't use update function. But now it works and config files gets saved. I will be on vacation the next 10 days. I will test the other functions to a limited amount bacause I won't have wifi and have to use my mobile plan.
Great job and thank you very much.

**Comment by outdoorbits** [comment]
Have a nice vacation, I try to present you more to test later... ;-)

**Reply by roshin8** [reply]
I would love to test out stuff from my end and help in any way possible with this :) i'm knowledgeable in python as well if that helps

**Comment by outdoorbits** [comment]
Warmly welcome!!!

**Comment by outdoorbits** [comment]
This discussion has now covered many different issues and has therefore become confusing.  Therefore, I hereby close it and open a new one to prepare the release of the new version on October 9th.
